### ISSUE TITLE
 import/no-useless-path-segments doesn't work on aliased project-relative imports #237 

### ISSUE THREAD

**Issue by icatalina** [import-js/eslint-plugin-import#2865]
When using an absolute import `import/no-useless-path-segments` is ignored:

## Absolute import:
<img width="410" alt="image" src="https://github.com/import-js/eslint-import-resolver-typescript/assets/4542735/79585bf2-fef6-4bbe-8557-03905b6af7d2">

## Relative import:
<img width="345" alt="image" src="https://github.com/import-js/eslint-import-resolver-typescript/assets/4542735/6e364622-78a6-48bc-9479-db2debdeae00">

**Comment by ljharb** [comment]
That's expected, because only on relative paths can those assumptions be safely applied.

**Comment by halcarleton** [comment]
> That's expected, because only on relative paths can those assumptions be safely applied.

@ljharb Could you expand on this? I do not see why this rule could not be safely applied to absolute paths. I'm also not seeing anything in the documentation of that rule explaining this.

**Comment by ljharb** [comment]
Because absolute paths aren't portable outside a project, which is why they should never be used regardless. See https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/no-absolute-path.md

**Comment by ljharb** [comment]
ah. Can you share your eslint config? Aliases like that need to have a custom resolver wired up to the settings.

**Comment by halcarleton** [comment]
> (ps those are called “project-relative”, the only absolute paths ate actually absolute ones, and next’s docs are wrong to call them that)

I get your point, but you are wrong. If the docs call them absolute imports then in the context of NextJs these are absolute imports.

**Comment by ljharb** [comment]
@halcarleton you're wrong, but it's fine, i've already reached out to vercel employees to fix their incorrect docs. so soon they won't be incorrectly calling them "absolute" anymore.

**Comment by icatalina** [comment]
My bad then, I was referring to _project-relative_ paths, not _absolute_ ones. I'd imagine an absolute path has to start with `/`, in my example they were all `src/`. 🤷

**Comment by ljharb** [comment]
@icatalina in that case i would definitely expect it to work.

What's your eslint config? Are you using the TS eslint parser and eslint-import-resolver-typescript?

**Comment by Arkellys** [comment]
Just to clarify since the conversation is a little confusing, are imports from aliases supposed to work with the rule? I don't get warned when I have wrong import paths such as:
```js
import x from "@app/something//foo/bar";
```

With `jsconfig.json`:

```json
{
  "compilerOptions": {
    ...
    "paths": {
      "@app/*": [
        "src/*"
      ],
    }
  },
}
```

**Comment by ljharb** [comment]
I wouldn't expect jsconfig.json to work (unless tsconfig-paths supports it in v3). You'd also need to have the typescript resolver configured.

**Comment by ljharb** [comment]
really you'd need a jsconfig resolver, but none exists that i'm aware of. jsconfig isn't really a standard thing, so few things support it.

**Comment by Arkellys** [comment]
So in conclusion, this rule doesn't work with aliases when not in a TS project?

**Comment by ljharb** [comment]
It works with multiple kinds of aliases - webpack ones, tsconfig ones, babel ones - but since `jsconfig` isn't any kind of standard, someone would have to create an `eslint-import-resolver-jsconfig` package.

and it looks like someone did! https://www.npmjs.com/package/eslint-import-resolver-jsconfig that should solve your issue.

**Comment by Arkellys** [comment]
I completely forgot that my aliases are also specified in my `.eslintrc`:

```js
settings: {
  "import/resolver": {
    alias: {
      map: [["@app", "./src"]]
    }
},
```

I use [`eslint-import-resolver-alias`](https://github.com/johvin/eslint-import-resolver-alias) for it.
Is the rule supposed to work in this case then? 😅

**Comment by ljharb** [comment]
ok well, using the eslint settings it should definitely work - i'm not sure if it's a problem here, or with that resolver tho?

(either way, whatever's doing the actual work so that those paths resolve at runtime in production, should probably be the resolver you use)

**Comment by Arkellys** [comment]
I can try another resolver when I have some time, maybe [this one](https://github.com/laysent/eslint-import-resolver-custom-alias) since the other one is archived, and see if it makes a difference!

**Comment by Arkellys** [comment]
It's not working with the other resolver either, so I don't think the problem comes from this.

**Comment by ljharb** [comment]
Typically I use babel, and a babel alias resolver, to do custom aliases - i'm not even sure what tooling supports jsconfig.

**Comment by Arkellys** [comment]
Sorry, I was probably not clear on my first message. What is not working for me is the linting on the IDE (via [VSCode ESLint extension](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint)). The aliases defined on the jsconfig are (in my case) only for VSCode IntelliSense, and I shouldn't have mentioned them. I use the plugins `eslint-import-resolver-alias` or `eslint-import-resolver-custom-alias` to handle the aliases for ESLint. Linting via the IDE should not require babel or webpack or other tools like this... or is there something I'm missing?

**Comment by ljharb** [comment]
No, you're totally right - but what is making the aliases work at runtime?

**Comment by Arkellys** [comment]
If I'm not wrong about the meaning of "runtime", then I use webpack [`resolve.alias`](https://webpack.js.org/configuration/resolve/#resolvealias). But since I don't use [EslintWebpackPlugin](https://webpack.js.org/plugins/eslint-webpack-plugin/), ESLint is never involved.

**Comment by ljharb** [comment]
ok - if you're using webpack to make the aliases work, then i'd suggest using the webpack resolver instead of the module-alias ones.

**Comment by Arkellys** [comment]
But my aliases already work fine with the ESLint resolvers, the problem don't come from them. And I don't see how is webpack related to all of this, again I'm talking about linting on the IDE, no code compilation involved. The rule just doesn't seem to support aliases given this configuration.

**Comment by ljharb** [comment]
@Arkellys the way this plugin works, it needs a resolver. You're 100% right that the resolver doesn't *have* to match the actual runtime mechanism that makes it work - but it's much simpler if it does, so I'd suggested trying the webpack resolver instead of the alias one.

"linting on the IDE doesn't work" and "the ESLint resolvers work" are conflicting statements - unless you're saying it works on the command line and not in the IDE, in which case it's an IDE-specific issue.

**Comment by Arkellys** [comment]
Neither the CLI nor linting on the IDE works. I'm saying the resolvers work because if I don't use one I will get lint errors on imports using aliases (via the rule `import/no-unresolved` if I remember correctly). I don't want to involve webpack into my ESLint stuff, I just wanted to confirm that it was a lack of support of `import/no-useless-path-segments`. Thank you.

**Comment by ljharb** [comment]
ahh, sorry, i'd forgotten the context here :-) thanks for bearing with me.

Indeed, see https://github.com/import-js/eslint-plugin-import/issues/2865#issuecomment-1692309170 - only paths that remain actually relative *in your source code* are supported by that rule.

Closing, since the original question was answered.
